### DISCUSSION TITLE
422: "value is not a valid dict" on Form() BaseModel validation - @app.post uses Path(), File() and Form() - Request called from SwaggerUI

### DISCUSSION THREAD

**Discussion by baudneo** [fastapi/fastapi#6304] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from enum import Enum

from fastapi import (
    FastAPI,
    File,
    UploadFile,
    Form,
)
from pydantic import BaseModel, Field

app = FastAPI()


# Models
class ModelProcessor(str, Enum):
    CPU = "cpu"
    GPU = "gpu"
    TPU = "tpu"
    DEFAULT = CPU


class ModelThresholds(BaseModel):
    confidence: float = Field(0.5, ge=0.0, le=1.0, descritpiton="Confidence Threshold")
    nms: float = Field(
        0.4, ge=0.0, le=1.0, description="Non-Maximum Suppression Threshold"
    )


class ModelOptions(BaseModel):
    processor: ModelProcessor = Field(None, description="Processor to use for model")
    height: int = Field(
        416, ge=1, description="Height of the input image (resized for model)"
    )
    width: int = Field(
        416, ge=1, description="Width of the input image (resized for model)"
    )
    square: bool = Field(False, description="Zero pad the image to be a square")
    fp_16: bool = Field(False, description="model uses Floating Point 16 Backend (EXPERIMENTAL!)")

    thresholds: ModelThresholds = Field(
        default_factory=ModelThresholds, description="Thresholds for the model"
    )


@app.post("/detect/{model_uuid}", summary="Run detection on an image")
async def object_detection(
        model_uuid: str,
        model_options: ModelOptions = Form(),
        image: UploadFile = File(..., description="Image to run the ML model on"),
):
    print(f"Running detection on image '{image.filename}'")
    return {"filename": image.filename}


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=5001)
```

### Description

# What I am doing
* Open browser and head to http://localhost:5001/docs to open the SwaggerUI
* Try out the only defined endpoint, you can enter any string into `model_uuid` field and any image for the `image: UploadFile`, the `model_options` is prefilled with proper JSON based upon the BaseModel.
* Hit execute, a '422: Unprocessable Entity' error is returned.
# What is happening
Error returned is 422: Unprocessable Entity ->
```
{
  "detail": [
    {
      "loc": [
        "body",
        "model_options"
      ],
      "msg": "value is not a valid dict",
      "type": "type_error.dict"
    }
  ]
}
```

# What I expect to happen
I expect that the return statement is returned with a 200 response code.

### Operating System

Linux

### FastAPI Version

0.79.0

### Python Version

3.10.6

### Additional Context

![Screenshot_2022-09-05_09-42-25](https://user-images.githubusercontent.com/86508179/188483628-4ddb60ae-9939-42f9-ace0-6b9abc6a3ad6.png)
![Screenshot_2022-09-05_09-42-33](https://user-images.githubusercontent.com/86508179/188483629-394f6935-c9ca-47dd-b48e-5073697d0648.png)

**Comment by iudeen** [comment]
I don't think we can use a nested Pydantic model with `Form()` in a straightforward way. I will verify anyway

**Comment by sheilalarmour** [comment]
Hi all.

**Comment by mbroton** [comment]
Hi! As described in documentation:
[Request Forms and Files](https://fastapi.tiangolo.com/tutorial/request-forms-and-files)

> You can declare multiple File and Form parameters in a path operation, but you can't also declare Body fields that you expect to receive as JSON, as the request will have the body encoded using multipart/form-data instead of application/json.
> 
> This is not a limitation of FastAPI, it's part of the HTTP protocol.

The workaround for this would be to send JSON as a string and parse it on your own. But I'm not sure if it's recommended.

**Comment by baudneo** [comment]
> Hi! As described in documentation:
> [https://fastapi.tiangolo.com/tutorial/request-forms-and-files/](url)
> 
> > You can declare multiple File and Form parameters in a path operation, but you can't also declare Body fields that you expect to receive as JSON, as the request will have the body encoded using multipart/form-data instead of application/json.
> > 
> > This is not a limitation of FastAPI, it's part of the HTTP protocol.
> 
> The workaround for this would be to send JSON as a string and parse it on your own. But I'm not sure if it's recommended.

Hi, I did see this in the docs but I was not certain of if that applied as in the same bit of docs it describes using Form() for JSON? Or maybe I am misunderstanding?

**Comment by iudeen** [comment]
The only limitation here is you will not be able to have nested JSONs within form.

```python
from enum import Enum
from typing import Type
import inspect

from fastapi import (
    FastAPI,
    File,
    UploadFile,
    Form, Depends,
)
from pydantic import BaseModel, Field

app = FastAPI()

def as_form(cls: Type[BaseModel]):
    new_parameters = []

    for field_name, model_field in cls.__fields__.items():
        model_field: ModelField  # type: ignore

        new_parameters.append(
             inspect.Parameter(
                 model_field.alias,
                 inspect.Parameter.POSITIONAL_ONLY,
                 default=Form(...) if not model_field.required else Form(model_field.default),
                 annotation=model_field.outer_type_,
             )
         )

    async def as_form_func(**data):
        return cls(**data)

    sig = inspect.signature(as_form_func)
    sig = sig.replace(parameters=new_parameters)
    as_form_func.__signature__ = sig  # type: ignore
    setattr(cls, 'as_form', as_form_func)
    return cls

# Models
class ModelProcessor(str, Enum):
    CPU = "cpu"
    GPU = "gpu"
    TPU = "tpu"
    DEFAULT = CPU


class ModelThresholds(BaseModel):
    confidence: float = Field(0.5, ge=0.0, le=1.0, descritpiton="Confidence Threshold")
    nms: float = Field(
        0.4, ge=0.0, le=1.0, description="Non-Maximum Suppression Threshold"
    )

class ModelOptions(BaseModel):
    processor: ModelProcessor = Form(None, description="Processor to use for model")
    height: int = Form(
        416, ge=1, description="Height of the input image (resized for model)"
    )
    width: int = Form(
        416, ge=1, description="Width of the input image (resized for model)"
    )
    square: bool = Form(False, description="Zero pad the image to be a square")
    fp_16: bool = Form(False, description="model uses Floating Point 16 Backend (EXPERIMENTAL!)")

    # thresholds: ModelThresholds = Form(...)



@app.post("/detect/{model_uuid}", summary="Run detection on an image")
async def object_detection(
        model_uuid: str,
        model_options: ModelOptions = Depends(),
        image: UploadFile = File(..., description="Image to run the ML model on"),
):
    print(f"Running detection on image '{image.filename}'")
    return {"filename": image.filename}


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=5001)
```

In the above code, I just commented out `thresholds` and changed how you have formed the `BaseModel`

**Comment by baudneo** [comment]
@iudeen - I am just following what the docs say about uploading an image with extra data -> https://fastapi.tiangolo.com/tutorial/request-forms-and-files/

Perhaps I am misunderstanding though?

**Comment by mbroton** [comment]
Examples in documentation have `str` type of `Form`. You've used `ModelOptions`.

You can use @iudeen's solution, but you can't nest your models.
Or send JSON as a string if you have to nest:
```py3
@app.post("/detect/{model_uuid}", summary="Run detection on an image")
async def object_detection(
        model_uuid: str,
        json_as_str: Json = Form(),  # from pydantic import Json
        image: UploadFile = File(..., description="Image to run the ML model on"),
):
    print(f"Running detection on image '{image.filename}'")
    return {"filename": image.filename}
```
which will result in this:
![image](https://user-images.githubusercontent.com/50829834/188489097-fcdc60d2-cad7-42ef-82c1-059bd2e63194.png)

**Comment by baudneo** [comment]
Thank you for that, I am just trying to understand my error. So Form() will only accept str input? pydantic Json will take JSON formatted as a str and load it into a python dict?

Meaning I would need to instantiate a ModelOptions object using ModelOptions(**json_as_str) and then pass it to whatever function? My only concern thereafter would be the Swagger docs being incomplete on what the json_as_str should contain since it wont be able to reference ModelOptions.

Maybe I can set ModelOptions as Query()? So the end user knows what the options and defaults are?

Sorry if this seems obvious, I am just trying to understand.

**Comment by mbroton** [comment]
> pydantic Json will take JSON formatted as a str and load it into a python dict?

Something like that.

I haven't tested it, but I guess you can do something like:
```py3
@app.post("/detect/{model_uuid}", summary="Run detection on an image")
async def object_detection(
        model_uuid: str,
        model_options: Json = Form(),
        image: UploadFile = File(..., description="Image to run the ML model on"),
):
    print(f"Running detection on image '{image.filename}'")
    options = ModelOptions.parse_obj(model_options)
    return {"filename": image.filename}
```
and `options` would be an instance of ModelOptions.

I can't provide answers for other questions, since I'm not sure.

**Comment by baudneo** [comment]
Thank you for the help, I am testing a few things out with the supplied options. Appreciate it very much @iudeen and @mbroton!

**Comment by iudeen** [comment]
@baudneo one last attempt to get as close as you want this to be. This handles everything.

```python
import json
from enum import Enum
from typing import Type, Dict, Optional
import inspect

from fastapi import (
    FastAPI,
    File,
    UploadFile,
    Form, Depends,
)

from dataclasses import dataclass, field

from pydantic import BaseModel

app = FastAPI()


# Models
class ModelProcessor(str, Enum):
    CPU = "cpu"
    GPU = "gpu"
    TPU = "tpu"
    DEFAULT = CPU

class ModelThresholds(BaseModel):
    confidence: float = Form(0.5, ge=0.0, le=1.0, descritpiton="Confidence Threshold")
    nms: float = Form(
        0.4, ge=0.0, le=1.0, description="Non-Maximum Suppression Threshold"
    )

@dataclass
class ModelOptions:
    processor: ModelProcessor = Form(None, description="Processor to use for model")
    height: int = Form(
        416, ge=1, description="Height of the input image (resized for model)"
    )
    width: int = Form(
        416, ge=1, description="Width of the input image (resized for model)"
    )
    square: bool = Form(False, description="Zero pad the image to be a square")
    fp_16: bool = Form(False, description="model uses Floating Point 16 Backend (EXPERIMENTAL!)")

    thresholds: Optional[str] = Form(ModelThresholds().json())


@app.post("/detect/{model_uuid}", summary="Run detection on an image")
async def object_detection(
        model_uuid: str,
        model_options: ModelOptions = Depends(),
        image: UploadFile = File(..., description="Image to run the ML model on"),
):
    threshold = ModelThresholds(**json.loads(model_options.thresholds))
    print(threshold.dict())
    print(f"Running detection on image '{image.filename}'")
    return {"filename": image.filename}


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=5001)
```
